### ISSUE TITLE
OrdinaryDiffEq performance degredation

### ISSUE THREAD

**Issue by ChrisRackauckas** [SciML/OrdinaryDiffEq.jl#1385]
Found because our automated benchmarking is finally back up and running and we see:

https://benchmarks.sciml.ai/html/NonStiffODE/linear_wpd.html

degraded to what's shown in the attachment.

[linear_wpd.zip](https://github.com/SciML/OrdinaryDiffEq.jl/files/6359049/linear_wpd.zip)

**Comment by ChrisRackauckas** [comment]
@YingboMa @chriselrod it would be good to dig in and see what's going on here.

**Comment by devmotion** [comment]
What was the problem? Was it fixed by https://github.com/SciML/OrdinaryDiffEq.jl/pull/1431 and https://github.com/SciML/OrdinaryDiffEq.jl/pull/1429? And are the DelayDiffEq regressions fixed as well by these changes?

**Comment by ChrisRackauckas** [comment]
Some chats on this. The solution was FastBroadcast.jl replacing the version in DiffEqBase which rotted a bit.


```
https://benchmarks.sciml.ai/html/NonStiffODE/linear_wpd.html
10:45
is what it was
10:45
that first picture
10:45
I like to look at that Tsit5 knob, at like 10^(2.4)
10:46
on AMDCI3: https://s3.amazonaws.com/buildkiteartifacts.com/d409823c-5fa7-41c8-9033-7269c5fde4f3/[…]65fe4f6d25d2d6718e3ef96dc3bde0f838106e3e82eac1a0ea9dde0b439f
10:47
(I don't think the s3 links are permanent ... does that work for you?
10:47
https://s3.amazonaws.com/buildkiteartifacts.com/d409823c-5fa7-41c8-9033-7269c5fde4f3/[…]d3c9e4876176d982b787ae6af4d4329e4380b8651c92050185b5dca1eb2d
10:47
and that's on the benchmarking gpuci1
10:47
all the way up to 10^(-2)
10:47
that computer is the same as what generated it before
10:48
so at first I thought it might've been an AMD thing, but no. (edited) 
10:48
and notice only the Julia ones moved, so that's a sign of a regression.
10:49
locally it's just
10:49
using OrdinaryDiffEq
Random.seed!(123)
# 2D Linear ODE
function f(du,u,p,t)
  @inbounds for i in eachindex(u)
    du[i] = 1.01*u[i]
  end
end
function f_analytic(u₀,p,t)
  u₀*exp(1.01*t)
end
tspan = (0.0,10.0)
prob = ODEProblem(ODEFunction(f,analytic=f_analytic),rand(100,100),tspan)
solve(prob,Tsit5(),save_everystep=false,abstol=1e-10,reltol=1e-7)
10:50
Capture.PNG 
Capture.PNG


10:50
The profile shows it's all in broadcasts
10:50
the circled part is calculate_residuals
10:51
seen in another way

Yingbo Ma  10:51 AM
There’s https://github.com/YingboMa/FastBroadcast.jl
YingboMa/FastBroadcast.jl
Stars
6
Language
Julia
Added by GitHub

1 reply
2 months agoView thread

Chris Rackauckas  10:51 AM
Capture.PNG 
Capture.PNG



Yingbo Ma  10:51 AM
but there are more cases to handle in FastBroadcast.jl

Chris Rackauckas  10:51 AM
this is all already using @..
10:51
but I think the piece that might have regressed is calculate_residuals

Yingbo Ma  10:52 AM
FastBroadcast.jl is different than @..

Chris Elrod  10:52 AM
One of its goals is to handle more cases.

Yingbo Ma  10:53 AM
We can have a mode where we disable [1.0] .+ [1, 2]

2 replies
Last reply 2 months agoView thread

Chris Rackauckas  10:53 AM
https://github.com/SciML/DiffEqBase.jl/blob/2daa81bfeb3df95c43469b96c058877a9cc240da/src/calculate_residuals.jl#L1-L12

calculate_residuals.jl
"""
    calculate_residuals(ũ, u₀, u₁, α, ρ, internalnorm, t)

Calculate element-wise residuals
```math
Show more
<https://github.com/SciML/DiffEqBase.jl|SciML/DiffEqBase.jl>SciML/DiffEqBase.jl | Added by GitHub
10:53
it seems like the max trips up SIMD but might not have before?

Chris Elrod  10:54 AM
Try Base.FastMath.max_fast
10:56
This SIMDs on 1.7:
julia> vmax!(z,x,y) = z .= max.(x,y)
vmax! (generic function with 1 method)

Chris Rackauckas  10:59 AM
Capture.PNG 
Capture.PNG


10:59
it's not as bad on my local machine though?

Chris Elrod  11:03 AM
It's a little better, but still looks much worse on the old ones.
11:03
What internalnorm should I use?
11:03
Maybe it isn't specializing on that when it once did?

Chris Rackauckas  11:05 AM
https://benchmarks.sciml.ai/html/NonStiffODE/linear_wpd.html shows the package versions it was using.
11:08
it should just be the default internalnorm

Yingbo Ma  11:14 AM
julia> @btime DiffEqBase.calculate_residuals!($out, $ũ, $u₀, $u₁, $α, $ρ, $internalnorm, $t);
  11.188 μs (0 allocations: 0 bytes)
julia>
julia> @btime DiffEqBase.calculate_residuals!($out, $ũ, $u₀, $u₁, $α, $ρ, $internalnorm, $t);
  20.854 μs (0 allocations: 0 bytes)
11:14
The first one is master DiffEqBase
11:14
the second is with @fastmath on the scalar calculate_residuals

Chris Elrod  11:15 AM
100% regression, interesting.

Chris Rackauckas  11:16 AM
Think it could be Julia Base?

Chris Elrod  11:16 AM
Have code to reproduce?
11:17
What is the default internalnorm?

Chris Rackauckas  11:17 AM
using OrdinaryDiffEq
Random.seed!(123)
# 2D Linear ODE
function f(du,u,p,t)
  @inbounds for i in eachindex(u)
    du[i] = 1.01*u[i]
  end
end
function f_analytic(u₀,p,t)
  u₀*exp(1.01*t)
end
tspan = (0.0,10.0)
prob = ODEProblem(ODEFunction(f,analytic=f_analytic),rand(100,100),tspan)
solve(prob,Tsit5(),save_everystep=false,abstol=1e-10,reltol=1e-7)
11:17
is the local reproducer
11:17
https://github.com/SciML/DiffEqBase.jl/blob/master/src/common_defaults.jl#L11-L18

common_defaults.jl
@inline ODE_DEFAULT_NORM(u::Union{AbstractFloat,Complex},t) = @fastmath abs(u)
@inline ODE_DEFAULT_NORM(u::Array{T},t) where T<:Union{AbstractFloat,Complex} =
                                         sqrt(real(sum(abs2,u)) / length(u))
@inline ODE_DEFAULT_NORM(u::StaticArrays.StaticArray{T},t) where T<:Union{AbstractFloat,Complex} =
                                            sqrt(real(sum(abs2,u)) / length(u))
Show more
<https://github.com/SciML/DiffEqBase.jl|SciML/DiffEqBase.jl>SciML/DiffEqBase.jl | Added by GitHub
11:17
that's the default norm
:+1:
1


Chris Elrod  11:18 AM
For calculate_residuals?

Yingbo Ma  11:18 AM
Oops, I mean the reverse
11:18
fastmath is faster
11:19
https://github.com/SciML/DiffEqBase.jl/pull/661
#661 Add fastmath on scalar `calculate_residuals`
using DiffEqBase
out, ũ, u₀, u₁ = [rand(100,100) for i in 1:4];
α, ρ, internalnorm, t = 1e-4, 1e-6, DiffEqBase.ODE_DEFAULT_NORM, 0.1;
DiffEqBase.calculate_residuals!(out, ũ, u₀, u₁, α, ρ, internalnorm, t);
using BenchmarkTools
Show more
<https://github.com/SciML/DiffEqBase.jl|SciML/DiffEqBase.jl>SciML/DiffEqBase.jl | May 3rd | Added by GitHub

Chris Elrod  11:21 AM
First without fastmath, second with, so about the same thing you saw:
julia> @btime DiffEqBase.calculate_residuals!($out, $ũ, $u₀, $u₁, $α, $ρ, $internalnorm, $t);
  22.295 μs (0 allocations: 0 bytes)
julia> @btime DiffEqBase.calculate_residuals!($out, $ũ, $u₀, $u₁, $α, $ρ, $internalnorm, $t);
  9.641 μs (0 allocations: 0 bytes)
(edited)
11:22
julia> @benchmark solve($prob,Tsit5(),save_everystep=false,abstol=1e-10,reltol=1e-7)
BenchmarkTools.Trial:
  memory estimate:  1.68 MiB
  allocs estimate:  76
  --------------
  minimum time:     10.348 ms (0.00% GC)
  median time:      10.497 ms (0.00% GC)
  mean time:        10.552 ms (0.25% GC)
  maximum time:     13.579 ms (22.44% GC)
  --------------
  samples:          474
  evals/sample:     1
julia> @benchmark solve($prob,Tsit5(),save_everystep=false,abstol=1e-10,reltol=1e-7)
BenchmarkTools.Trial:
  memory estimate:  1.68 MiB
  allocs estimate:  76
  --------------
  minimum time:     9.723 ms (0.00% GC)
  median time:      9.864 ms (0.00% GC)
  mean time:        9.908 ms (0.23% GC)
  maximum time:     11.407 ms (11.51% GC)
  --------------
  samples:          505
  evals/sample:     1
First without fastmath, second with, so a small improvement.

Yingbo Ma  11:22 AM
It’s still not vectorizing calculate_residuals

Chris Elrod  11:23 AM
Yeah, just a bunch of "sd" instructions

Chris Rackauckas  11:23 AM
using OrdinaryDiffEq, Sundials, DiffEqDevTools, Plots, ODEInterfaceDiffEq, ODE, LSODA
using Random
Random.seed!(123)
gr()
# 2D Linear ODE
function f(du,u,p,t)
  @inbounds for i in eachindex(u)
    du[i] = 1.01*u[i]
  end
end
function f_analytic(u₀,p,t)
  u₀*exp(1.01*t)
end
tspan = (0.0,10.0)
prob = ODEProblem(ODEFunction(f,analytic=f_analytic),rand(100,100),tspan)
solve(prob,Tsit5(),save_everystep=false,abstol=1e-10,reltol=1e-7)
abstols = 1.0 ./ 10.0 .^ (3:13)
reltols = 1.0 ./ 10.0 .^ (0:10);
setups = [Dict(:alg=>DP5())
          Dict(:alg=>ode45())
          Dict(:alg=>dopri5())
          Dict(:alg=>ARKODE(Sundials.Explicit(),etable=Sundials.DORMAND_PRINCE_7_4_5))
          Dict(:alg=>Tsit5())]
solnames = ["OrdinaryDiffEq";"ODE";"ODEInterface";"Sundials ARKODE";"OrdinaryDiffEq Tsit5"]
wp = WorkPrecisionSet(prob,abstols,reltols,setups;names=solnames,save_everystep=false,numruns=100)
plot(wp)
is the full one to generate the plot BTW.
11:23
I'm going to fix that Sundials issue in the meantime :sweat_smile:

Yingbo Ma  11:25 AM
@inline function calculate_residuals!(out, ũ, u₀, u₁, α, ρ, internalnorm,t)
    @simd ivdep for ii in eachindex(out, ũ, u₀, u₁)
        @inbounds out[ii] = calculate_residuals(ũ[ii], u₀[ii], u₁[ii], α, ρ, internalnorm,t)
    end
    nothing
end
gives
julia> @btime DiffEqBase.calculate_residuals!($out, $ũ, $u₀, $u₁, $α, $ρ, $internalnorm, $t);
  5.575 μs (0 allocations: 0 bytes)
11:25
It’s just broadcasting sucks

Chris Rackauckas  11:27 AM
Did it work in Julia v1.4?

Yingbo Ma  11:27 AM
I don’t have 1.4 in handy

Chris Rackauckas  11:28 AM
should we just specialize this on ::Array?

Yingbo Ma  11:28 AM
but let’s just fix this anyway
11:28
FastBroadcast specializes ::Array if you want to wait for a “proper” fix

Chris Elrod  11:28 AM
We should also be able to fix it with FastBroadcast.jl
11:29
It doesn't use the broadcast machinery, and instead builds a proper loop like Fortran intended
:boom:
1
:fire:
1


Yingbo Ma  11:30 AM
tbh, I don’t get why Julia doesn’t do that
11:30
It compiles faster, too.

Chris Elrod  11:31 AM
"@generated is a hack"
11:32
Generated functions aren't easy to read, but many layers of dispatch aren't easy to follow either.
11:34
Is the point of using broadcasts here for readability, or for being generic with respect to something being a scalar or vector?
11:34
If the latter, we could just write loops using something like:
@inline _maybeindex(x::Number, i) = x
@inline _maybeindex(x::AbstractArray, i) = @inbounds x[i]
11:38
julia> @btime $out .= $ũ ./ $u₀;
  5.223 μs (0 allocations: 0 bytes)
julia> @btime DiffEqBase.calculate_residuals!($out, $ũ, $u₀, $u₁, $α, $ρ, $internalnorm, $t);
  4.686 μs (0 allocations: 0 bytes)

Chris Rackauckas  11:38 AM
https://github.com/SciML/Sundials.jl/pull/304 Makes Sundials a bit less annoying in this benchmark.

#304 only set the nonlinear iters if there is a nonlinear solver
<https://github.com/SciML/Sundials.jl|SciML/Sundials.jl>SciML/Sundials.jl | May 3rd | Added by GitHub

Chris Elrod  11:56 AM
At this point, the division is the bottleneck if we want to optimize calculate_residuals further. If we want to go that route, I would replace some (but not all) divisions with an approximate inverse + a few NR iterations. Implementing that in a hacky way:
julia> @btime DiffEqBase.calculate_residuals!($out, $ũ, $u₀, $u₁, $α, $ρ, $internalnorm, $t); # 100% divisions
  4.687 μs (0 allocations: 0 bytes)
julia> @btime calculate_residuals_avx_loop!($out, $ũ, $u₀, $u₁, $α, $ρ, $internalnorm, $t); # 75% divisions + 25% approximate inverse + 2 NR iterations
  3.540 μs (0 allocations: 0 bytes)
The idea there is that the floating point divider is a separate unit that can work in parallel to the others, but it is very slow. This means that if you do just a few divisions, they're "free" in that they don't hog resources from anything else. Do more than a few, and it easily becomes the bottleneck of the program. So ideally you'd balance the cost by taking advantage of the fact that an approximate inverse with NR iterations doesn't use the divider, and with 2 iterations can be accurate to within just a couple ULP.
But this probably only works well on AVX512, because without it you'd need to convert to Float32 and back, and the speed of div is less of an issue with 256-bit vector CPUs anyway, especially Zen2 and Zen3, which have pretty fast SIMD divisions (compared to Intel, still slow compared to other stuff) (edited) 
11:57
With the regular @simd loop:
julia> @benchmark solve($prob,Tsit5(),save_everystep=false,abstol=1e-10,reltol=1e-7)
BenchmarkTools.Trial:
  memory estimate:  1.68 MiB
  allocs estimate:  76
  --------------
  minimum time:     9.374 ms (0.00% GC)
  median time:      9.521 ms (0.00% GC)
  mean time:        9.570 ms (0.32% GC)
  maximum time:     11.506 ms (15.29% GC)
  --------------
  samples:          523
  evals/sample:     1

Chris Elrod  12:05 PM
>85% of the time is being spent in perform_step! (edited) 

Yingbo Ma  12:06 PM
Yeah, that’s also many broadcasts

Chris Rackauckas  12:09 PM
those should already be pretty optimal, right?

Chris Elrod  12:09 PM
https://github.com/SciML/OrdinaryDiffEq.jl/blob/41f39e4d44ea7271d683c4f6142775b9204130b6/src/perform_step/low_order_rk_perform_step.jl#L627
low_order_rk_perform_step.jl
@muladd function perform_step!(integrator, cache::Tsit5Cache, repeat_step=false)
<https://github.com/SciML/OrdinaryDiffEq.jl|SciML/OrdinaryDiffEq.jl>SciML/OrdinaryDiffEq.jl | Added by GitHub

Chris Elrod  12:20 PM
julia> @benchmark solve($prob,Tsit5(),save_everystep=false,abstol=1e-10,reltol=1e-7)
BenchmarkTools.Trial:
  memory estimate:  1.68 MiB
  allocs estimate:  72
  --------------
  minimum time:     3.883 ms (0.00% GC)
  median time:      4.168 ms (0.00% GC)
  mean time:        4.222 ms (0.69% GC)
  maximum time:     6.738 ms (34.27% GC)
  --------------
  samples:          1184
  evals/sample:     1
12:20
After replacing all of them with loops
12:20
@Yingbo Ma yeah, we definitely should finish up FastBroadcast and make sure it can be a drop in replacement for the current @..

Chris Rackauckas  12:21 PM
interesting
12:21
yeah it seems like a regression in @.. then?
12:21
Because we tested that it was as fast as loops before.

Chris Elrod  12:21 PM
Seems so. @inbounds @simd is much faster now.
12:22
And is actually SIMDing. I didn't look at the asm here since I thought it'd be as much work to dig my way here as it would be to just change it and benchmark.
12:23
But in the other example, @.. was generating a bunch of garbage asm (over 2x the loop version) without unrolling or SIMD, while the `@simd` loop generated lots of the good code and much less boiler plate. (edited) 

Yingbo Ma  12:23 PM
Calling perform_step is actually pretty ez
12:23
You can get the integrator by
integ = init(prob,Tsit5(),save_everystep=false,abstol=1e-10,reltol=1e-7)
12:23
and the cache is integ.cache
12:24
So it’s something like OrdinaryDiffEq.perform_step!(integ, integ.cache, false)

Chris Rackauckas  12:27 PM
okay, so is the solution going to be a full replacement to FastBroadcast.jl?

Yingbo Ma  12:29 PM
yeah

Chris Elrod  12:30 PM
Okay, just looked at perform_step! and the broadcast version was well over 2x the code. I counted 13 vs 7 SIMD loops in the loop vs broadcast versions.

Chris Rackauckas  12:58 PM
Does FastBroadcast reduce compile time too?
12:59
That would be significant to Pumas

Yingbo Ma  1:02 PM
yes, it tries to lower to normal loops as much as possible
1:03
Basically, FastBroadcast is trying to not do broadcast at all.

Chris Elrod  1:09 PM
Do we want to be clever and swap in SIMD-able versions of functions?

Chris Rackauckas  1:12 PM
I would say yes
:+1:
1


Chris Elrod  1:18 PM
I need to factor apart the JuliaSIMD packages more.
I think ideally SLEEFPirates shouldn't have to depend on VectorizationBase, which has gotten rather heavy as a dependency (1.5s load time). Plus, the more things can be made to not depend on one another, the more they can precompile in parallel. (edited) 
1:19
Using the functions from SLEEFPirates would be the easiest way.
```
